lib/ansible/vars/manager.py

```diff
@@ -383,19 +387,21 @@ def plugins_by_groups():
                 raise AnsibleParserError("Error while reading vars files - please supply a list of file names. "
                                          "Got '%s' of type %s" % (vars_files, type(vars_files)))
 
-            # By default, we now merge in all vars from all roles in the play,
+            # By default, we now merge in all exported vars from all roles in the play,
             # unless the user has disabled this via a config option
             if not C.DEFAULT_PRIVATE_ROLE_VARS:
+                # TODO: check per role privacy?
                 for role in play.get_roles():
-                    all_vars = _combine_and_track(all_vars, role.get_vars(include_params=False), "role '%s' vars" % role.name)
+                    if not role.from_include or role.public or (role.static and role._completed.get(to_text(host), False)):
+                        all_vars = _combine_and_track(all_vars, role.get_vars(include_params=False, only_exports=True), "role '%s' exported vars" % role.name)
 
         # next, we merge in the vars from the role, which will specifically
         # follow the role dependency chain, and then we merge in the tasks
         # vars (which will look at parent blocks/task includes)
         if task:
             if task._role:
-                all_vars = _combine_and_track(all_vars, task._role.get_vars(task.get_dep_chain(), include_params=False),
-                                              "role '%s' vars" % task._role.name)
+                all_vars = _combine_and_track(all_vars, task._role.get_vars(task.get_dep_chain(), include_params=True, only_exports=False),
```

Comment:
I see the comment below that the duplicate below needs to be removed for this to have effect, but I don't understand what the precedence change will be fixing. Role `params` having higher precedence than `include_vars` matches the docs https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable.

Reply:
Isn't it different from the docs now that you removed it? It seemed closer to the docs before.

Reply:
the big difference is that the precedence inside role stays the same but that the 'vars:'  and include_params are not exported as the 'vars/main.yml' is  so the 'general' precedence is affected since they are 'not in scope' outside the role.

Reply:
Is it difficult to add test coverage for that? The several test suites modified pass without the precedence change too.

Reply:
yep, that should be my next step